### PR TITLE
Update HttpRequestNode

### PR DESCRIPTION
**Pull request by Zluurk**
post and put, get params, test

### CODE REVIEW

File: pypeman/contrib/http.py [hunk 1]
```diff
@@ -128,23 +129,50 @@ def generate_request_url(self, msg):
...
-            conn = aiohttp.TCPConnector(verify_ssl=self.verify)
+            conn = aiohttp.TCPConnector(verify_ssl=self.verify, loop=loop)
 
         headers = self.headers
         if not headers:
             headers = msg.meta.get('headers')
         method=self.method
         if not method:
             method = msg.meta.get('method','get')
+        params=self.params
```
Comment (jrmi) [inline pypeman/contrib/http.py]: Missing space around assignment

File: pypeman/contrib/http.py [hunk 2]
```diff
@@ -128,23 +129,50 @@ def generate_request_url(self, msg):
             else:
                 ssl_context = ssl.SSLContext(ssl.PROTOCOL_SSLv23)
             ssl_context.load_cert_chain(self.client_cert[0], self.client_cert[1])
-            conn = aiohttp.TCPConnector(ssl_context=ssl_context)
+            conn = aiohttp.TCPConnector(ssl_context=ssl_context, loop=loop)
         else:
-            conn = aiohttp.TCPConnector(verify_ssl=self.verify)
+            conn = aiohttp.TCPConnector(verify_ssl=self.verify, loop=loop)
 
         headers = self.headers
```
Comment (jrmi) [inline pypeman/contrib/http.py]: Use `first_not_none` function here.

File: pypeman/contrib/http.py [hunk 3]
```diff
@@ -128,23 +129,50 @@ def generate_request_url(self, msg):
...
         method=self.method
         if not method:
             method = msg.meta.get('method','get')
+        params=self.params
+        if not params:
+            params = msg.meta.get('params', None)
+
+        get_params = []
+        if params:
+            for key in iter(params):
```
Comment (jrmi) [inline pypeman/contrib/http.py]: Can you use:

```python
for key, param in params.items():
    ...
```

File: pypeman/contrib/http.py [hunk 4]
```diff
@@ -128,23 +129,50 @@ def generate_request_url(self, msg):
...
+                if type(params[key]) is list:
```
Comment (jrmi) [inline pypeman/contrib/http.py]: It's more idiomatic to write:
```python
if isinstance(param, list):
    ...
```

File: pypeman/contrib/http.py [hunk 5]
```diff
@@ -128,23 +129,50 @@ def generate_request_url(self, msg):
...
+                    for value in params[key]:
+                        get_params.append((key, value))
+                else:
+                    get_params.append((key, params[key]))
+        else:
+            get_params = None
```
Comment (jrmi) [inline pypeman/contrib/http.py]: Can you inverse the logic here:
```python
get_params = None
if params:
     get_params = []
     ....
```

File: pypeman/contrib/http.py [hunk 6]
```diff
@@ -128,23 +129,50 @@ def generate_request_url(self, msg):
...
+
+
         if type(self.auth) == tuple:
```
Comment (jrmi) [inline pypeman/contrib/http.py]: Can you use `Isinstance` ?

File: pypeman/contrib/http.py [hunk 7]
```diff
@@ -128,23 +129,50 @@ def generate_request_url(self, msg):
...
             basic_auth = aiohttp.BasicAuth(self.auth[0], self.auth[1])
         else:
             basic_auth = self.auth
+
+        data=None
+        if method in ['put', 'post']:
+            data=msg.payload
```
Comment (jrmi) [inline pypeman/contrib/http.py]: Add space around assignment.

File: pypeman/tests/test_nodes.py [hunk 8]
```diff
@@ -287,6 +296,95 @@ def test_ftp_nodes(self):
...
+        channel = FakeChannel(self.loop)
+
+        auth = ("login", "mdp")
+        url = 'http://url/%(meta.beta)s/%(payload.alpha)s'
+        b_auth = aiohttp.BasicAuth(auth[0], auth[1])
+        client_cert = ('/cert.key', '/cert.crt')
+        http_node1 = nodes.HttpRequest(url=url, verify=False, auth=auth)
+        http_node1.channel = channel
+
+
```
Comment (jrmi) [inline pypeman/tests/test_nodes.py]: Too many empty lines.

File: pypeman/tests/test_nodes.py [hunk 9]
```diff
@@ -287,6 +296,95 @@ def test_ftp_nodes(self):
...
+        content1 = {"alpha": "payload_url"}
+        msg1 = generate_msg(message_content=content1)
+        meta_params = {'omega': 'meta_params'}
+        headers1 = {'test': 'test'}
+        msg1.meta = {"beta": "meta_url", 'params': meta_params, 'headers': headers1}
+        req_url1 = 'http://url/meta_url/payload_url'
+        req_kwargs1 = {
+            'data':None,
+            'params':[('omega', 'meta_params')],
+            'url':req_url1,
```
Comment (jrmi) [inline pypeman/tests/test_nodes.py]: Add one space after `:`

File: pypeman/tests/test_nodes.py [hunk 10]
```diff
@@ -287,6 +296,95 @@ def test_ftp_nodes(self):
...
+
+        with mock.patch('pypeman.contrib.http.aiohttp.ClientSession',
+        autospec=True) as mock_client_session, mock.patch('ssl.SSLContext',
+        autospec=True) as mock_ssl_context:
+            mock_ctx_mgr = mock_client_session.return_value
+            mock_session = mock_ctx_mgr.__enter__.return_value
+            mock_session.request = get_mock_coro(mock.MagicMock())
+            mock_load_cert_chain = mock_ssl_context.return_value.load_cert_chain
+
+            result = self.loop.run_until_complete(http_node1.handle(msg1))
```
Comment (jrmi) [inline pypeman/tests/test_nodes.py]: Can you add comment to explain what you're testing each step ?

File: pypeman/contrib/http.py [hunk 11]
```diff
@@ -92,14 +92,15 @@ def __init__(self, *args, endpoint=None, method='*', url='/', encoding=None, **k
 class HttpRequest(nodes.BaseNode):
     """ Http request node """
```
Comment (jrmi) [inline pypeman/contrib/http.py]: Can you enhance documentation to explain node params ?

File: pypeman/tests/test_nodes.py (hunk 10 above)
Comment (Zluurk) [reply pypeman/tests/test_nodes.py]: better ?

File: pypeman/tests/test_nodes.py (hunk 9 above)
Comment (Zluurk) [reply pypeman/tests/test_nodes.py]: ok

File: pypeman/tests/test_nodes.py (hunk 8 above)
Comment (Zluurk) [reply pypeman/tests/test_nodes.py]: ok

File: pypeman/contrib/http.py (hunk 7 above)
Comment (Zluurk) [reply pypeman/contrib/http.py]: ok

File: pypeman/contrib/http.py (hunk 6 above)
Comment (Zluurk) [reply pypeman/contrib/http.py]: ok

File: pypeman/contrib/http.py (hunk 5 above)
Comment (Zluurk) [reply pypeman/contrib/http.py]: ok

File: pypeman/contrib/http.py (hunk 4 above)
Comment (Zluurk) [reply pypeman/contrib/http.py]: ok

File: pypeman/contrib/http.py (hunk 1 above)
Comment (Zluurk) [reply pypeman/contrib/http.py]: ok

File: pypeman/contrib/http.py (hunk 11 above)
Comment (Zluurk) [reply pypeman/contrib/http.py]: better ?

File: pypeman/contrib/http.py (hunk 2 above)
Comment (Zluurk) [reply pypeman/contrib/http.py]: ok

**Pull request merged**